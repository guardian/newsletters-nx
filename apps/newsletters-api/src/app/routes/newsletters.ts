import type { FastifyInstance } from 'fastify';
import { transformDataToLegacyNewsletter } from '@newsletters-nx/newsletters-data-client';
import { newsletterStore } from '../../services/storage';
import { makeErrorResponse, makeSuccessResponse } from '../responses';

export function registerNewsletterRoutes(app: FastifyInstance) {
	// not using the makeSuccess function on this route as
	// we are emulating the response of the legacy API
	app.get('/api/legacy/newsletters', async (req, res) => {
		const storageResponse = await newsletterStore.list();
		if (!storageResponse.ok) {
			return res.status(500).send(makeErrorResponse(storageResponse.message));
		}

		return storageResponse.data.map(transformDataToLegacyNewsletter);
	});

	app.get('/api/newsletters', async (req, res) => {
		const storageResponse = await newsletterStore.list();
		if (!storageResponse.ok) {
			return res.status(500).send(makeErrorResponse(storageResponse.message));
		}

		return makeSuccessResponse(storageResponse.data);
	});

	app.get<{ Params: { newsletterId: string } }>(
		'/api/newsletters/:newsletterId',
		async (req, res) => {
			const { newsletterId } = req.params;

			// TO DO - newsletterStore.readByName
			const storageResponse = await newsletterStore.list();
			if (!storageResponse.ok) {
				return res.status(500).send(makeErrorResponse(storageResponse.message));
			}

			const newsletter = storageResponse.data.find(
				(newsletter) => newsletter.identityName === newsletterId,
			);

			if (!newsletter) {
				return res
					.status(404)
					.send(makeErrorResponse(`no match for id ${newsletterId}`));
			}

			return makeSuccessResponse(newsletter);
		},
	);
}
