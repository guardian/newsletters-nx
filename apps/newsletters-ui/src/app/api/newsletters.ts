import type {
	ApiResponse,
	Newsletter,
} from '@newsletters-nx/newsletters-data-client';

export async function getNewsletters(): Promise<Newsletter[]> {
	try {
		const response = await fetch('api/v1/newsletters');
		const data = (await response.json()) as ApiResponse<Newsletter[]>;

		return data.ok ? data.data : []; // TODO: handle non 2xx responses
	} catch (err) {
		console.error(err);
		return [];
	}
}

export async function getNewsletter(
	id: string,
): Promise<Newsletter | undefined> {
	try {
		const response = await fetch(`api/v1/newsletters/${id}`);
		const data = (await response.json()) as ApiResponse<Newsletter>;
		return data.ok ? data.data : undefined; // TODO: handle non 2xx responses
	} catch (err) {
		console.error(err);
		return undefined;
	}
}
