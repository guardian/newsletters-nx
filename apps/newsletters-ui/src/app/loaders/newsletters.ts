import type { LoaderFunction } from 'react-router';
import type {
	ApiResponse,
	Draft,
	NewsletterData,
} from '@newsletters-nx/newsletters-data-client';

async function fetchApiData<T>(path: string): Promise<T | undefined> {
	try {
		const response = await fetch(path);
		const data = (await response.json()) as ApiResponse<T>;
		return data.ok ? data.data : undefined;
	} catch (err) {
		console.error(err);
		return undefined;
	}
}

export const listLoader: LoaderFunction = async (): Promise<
	NewsletterData[]
> => {
	const list = (await fetchApiData<NewsletterData[]>(`api/newsletters`)) ?? [];
	return list;
};

export const detailLoader: LoaderFunction = async ({
	params,
}): Promise<NewsletterData | undefined> => {
	const { id } = params;
	if (!id) {
		return undefined;
	}
	return await fetchApiData<NewsletterData>(`api/newsletters/${id}`);
};

export const draftListLoader: LoaderFunction = async (): Promise<Draft[]> => {
	const list = (await fetchApiData<Draft[]>(`api/drafts`)) ?? [];
	return list;
};

export const draftDetailLoader: LoaderFunction = async ({
	params,
}): Promise<Draft | undefined> => {
	const { id } = params;
	if (!id) {
		return undefined;
	}
	return await fetchApiData<Draft>(`api/drafts/${id}`);
};
