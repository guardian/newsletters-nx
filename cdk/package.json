{
  "name": "cdk",
  "version": "0.0.0",
  "private": true,
  "scripts": {
    "build": "tsc",
    "test": "jest",
    "test-update": "jest -u",
    "format": "prettier --write \"{lib,bin}/**/*.ts\"",
    "lint": "eslint lib/** bin/** --ext .ts --no-error-on-unmatched-pattern",
    "synth": "cdk synth --path-metadata false --version-reporting false",
    "diff": "cdk diff --path-metadata false --version-reporting false"
  },
  "devDependencies": {
    "@guardian/cdk": "49.0.3",
    "@guardian/eslint-config-typescript": "3.0.0",
    "@guardian/prettier": "1.0.0",
    "@types/jest": "^29.2.5",
    "@types/node": "18.11.18",
    "aws-cdk": "2.59.0",
    "aws-cdk-lib": "2.59.0",
    "constructs": "10.1.215",
    "eslint": "^8.32.0",
    "jest": "^29.3.1",
    "prettier": "^2.8.3",
    "source-map-support": "^0.5.20",
    "ts-jest": "^29.0.5",
    "ts-node": "^10.9.1",
    "typescript": "~4.9.4"
  },
  "prettier": "@guardian/prettier",
  "jest": {
    "testMatch": [
      "<rootDir>/lib/**/*.test.ts"
    ],
    "transform": {
      "^.+\\.tsx?$": "ts-jest"
    },
    "setupFilesAfterEnv": [
      "./jest.setup.js"
    ]
  },
  "eslintConfig": {
    "root": true,
    "env": {
      "node": true,
      "jest": true
    },
    "extends": [
      "@guardian/eslint-config-typescript"
    ],
    "parserOptions": {
      "ecmaVersion": 2020,
      "sourceType": "module"
    },
    "plugins": [
      "@typescript-eslint"
    ],
    "rules": {
      "@typescript-eslint/no-inferrable-types": 0,
      "import/no-namespace": 2
    },
    "ignorePatterns": [
      "**/*.js",
      "node_modules",
      "cdk.out",
      ".eslintrc.js",
      "jest.config.js"
    ]
  }
}