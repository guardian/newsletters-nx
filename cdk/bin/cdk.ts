import 'source-map-support/register';
import { App } from 'aws-cdk-lib';
import { Newsletters } from '../lib/newsletters';
import { NewslettersApi } from '../lib/newsletters-api';

const app = new App();

// Shared props for all CDK apps
export const sharedProps = {
	stack: 'newsletters',
	env: {
		region: 'eu-west-1',
	},
};

/** The internal tool for newsletter management */
const newslettersAppName = 'newsletters';

new Newsletters(app, 'Newsletters-CODE', {
	...sharedProps,
	stage: 'CODE',
	app: newslettersAppName,
	domainName: `${newslettersAppName}.code.dev-gutools.co.uk`,
});

// new NewslettersApi(app, 'NewslettersUi-PROD', {
// 	...sharedProps,
// 	stage: 'PROD',
// 	app: newslettersUiAppName,
// 	domainName: `${newslettersUiAppName}.gutools.co.uk`,
// });

/** The read-only newsletters API */
const newslettersApiAppName = 'newsletters-api';

new NewslettersApi(app, 'NewslettersApi-CODE', {
	...sharedProps,
	stage: 'CODE',
	app: newslettersApiAppName,
	/** @todo - this ought to be a .guardianapis.co.uk domain */
	domainName: `${newslettersApiAppName}.code.dev-gutools.co.uk`,
});

// new NewslettersApi(app, 'NewslettersApi-PROD', {
// 	...sharedProps,
// 	stage: 'PROD',
// 	app: newslettersApiAppName,
// /** @todo - this ought to be a .guardianapis.co.uk domain */
// 	domainName: `${newslettersApiAppName}.gutools.co.uk`,
// });
