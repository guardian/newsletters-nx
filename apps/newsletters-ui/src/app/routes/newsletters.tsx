import type { RouteObject } from 'react-router-dom';
import { NewsletterDetailView } from '../components/views/NewsletterDetailView';
import { NewsletterEditView } from '../components/views/NewsletterEditView';
import { NewsletterJsonEditView } from '../components/views/NewsletterJsonEditView';
import { NewslettersListView } from '../components/views/NewslettersListView';
import { PreviewView } from '../components/views/PreviewView';
import { RenderingOptionsView } from '../components/views/RenderingOptionsView';
import { WizardContainer } from '../components/WizardContainer';
import { ErrorPage } from '../ErrorPage';
import { Layout } from '../Layout';
import { detailLoader, listLoader } from '../loaders/newsletters';

export const newslettersRoute: RouteObject = {
	path: '/newsletters',
	element: <Layout />,
	errorElement: <ErrorPage />,
	children: [
		{
			path: '',
			element: <NewslettersListView />,
			loader: listLoader,
		},
		{
			path: ':id',
			element: <NewsletterDetailView />,
			loader: detailLoader,
		},
		{
			path: 'edit/:id',
			element: <NewsletterEditView />,
			loader: detailLoader,
		},
		{
			path: 'rendering-options/:id',
			element: <RenderingOptionsView />,
			loader: detailLoader,
		},
		{
			path: 'edit-json/:id',
			element: <NewsletterJsonEditView />,
			loader: detailLoader,
		},
		{
			path: 'newsletter-data/:listId',
			element: <WizardContainer wizardId="NEWSLETTER_DATA" />,
		},
		{
			path: 'newsletter-data-rendering/:listId',
			element: <WizardContainer wizardId="RENDERING_OPTIONS" />,
		},
		{
			path: 'newsletter-data',
			element: <WizardContainer wizardId="NEWSLETTER_DATA" />,
		},
		{
			path: 'launch-newsletter/:listId',
			element: <WizardContainer wizardId="LAUNCH_NEWSLETTER" />,
		},
		{
			path: 'launch-newsletter',
			element: <WizardContainer wizardId="LAUNCH_NEWSLETTER" />,
		},

		{
			path: 'preview/:id',
			element: <PreviewView />,
			loader: detailLoader,
		},
	],
};
