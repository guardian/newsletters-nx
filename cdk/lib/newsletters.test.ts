import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { sharedProps } from '../bin/cdk';
import { Newsletters } from './newsletters';

describe('The Newsletters stack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const stack = new Newsletters(app, 'Newsletters-TEST', {
			...sharedProps,
			stage: 'TEST',
			app: 'newsletters',
			domainName: 'newsletters.test.dev-gutools.co.uk',
		});
		const template = Template.fromStack(stack);
		expect(template.toJSON()).toMatchSnapshot();
	});
});
