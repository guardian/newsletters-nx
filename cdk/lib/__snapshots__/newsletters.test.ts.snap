// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Newsletters stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuVpcParameter",
      "GuSubnetListParameter",
      "GuSubnetListParameter",
      "GuNodeApp",
      "GuCertificate",
      "GuInstanceRole",
      "GuDescribeEC2Policy",
      "GuLoggingStreamNameParameter",
      "GuLogShippingPolicy",
      "GuGetDistributablePolicy",
      "GuParameterStoreReadPolicy",
      "GuAmiParameter",
      "GuHttpsEgressSecurityGroup",
      "GuAutoScalingGroup",
      "GuWazuhAccess",
      "GuApplicationLoadBalancer",
      "GuApplicationTargetGroup",
      "GuHttpsApplicationListener",
      "GuStringParameter",
      "GuCname",
    ],
    "gu:cdk:version": "TEST",
  },
  "Outputs": {
    "LoadBalancerNewslettersDnsName": {
      "Description": "DNS entry for LoadBalancerNewsletters",
      "Value": {
        "Fn::GetAtt": [
          "LoadBalancerNewsletters7869FC55",
          "DNSName",
        ],
      },
    },
  },
  "Parameters": {
    "AMINewsletters": {
      "Description": "Amazon Machine Image ID for the app newsletters. Use this in conjunction with AMIgo to keep AMIs up to date.",
      "Type": "AWS::EC2::Image::Id",
    },
    "ClientId": {
      "Description": "Google OAuth client ID",
      "Type": "String",
    },
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "LoggingStreamName": {
      "Default": "/account/services/logging.stream.name",
      "Description": "SSM parameter containing the Name (not ARN) on the kinesis stream",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "VpcId": {
      "Default": "/account/vpc/primary/id",
      "Description": "Virtual Private Cloud to run EC2 instances within. Should NOT be the account default VPC.",
      "Type": "AWS::SSM::Parameter::Value<AWS::EC2::VPC::Id>",
    },
    "newslettersPrivateSubnets": {
      "Default": "/account/vpc/primary/subnets/private",
      "Description": "A list of private subnets",
      "Type": "AWS::SSM::Parameter::Value<List<AWS::EC2::Subnet::Id>>",
    },
    "newslettersPublicSubnets": {
      "Default": "/account/vpc/primary/subnets/public",
      "Description": "A list of public subnets",
      "Type": "AWS::SSM::Parameter::Value<List<AWS::EC2::Subnet::Id>>",
    },
  },
  "Resources": {
    "AutoScalingGroupNewslettersASG15647A78": {
      "Properties": {
        "HealthCheckGracePeriod": 120,
        "HealthCheckType": "ELB",
        "LaunchConfigurationName": {
          "Ref": "AutoScalingGroupNewslettersLaunchConfigE2413199",
        },
        "MaxSize": "2",
        "MinSize": "1",
        "Tags": [
          {
            "Key": "App",
            "PropagateAtLaunch": true,
            "Value": "newsletters",
          },
          {
            "Key": "gu:cdk:version",
            "PropagateAtLaunch": true,
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "PropagateAtLaunch": true,
            "Value": "guardian/newsletters-nx",
          },
          {
            "Key": "LogKinesisStreamName",
            "PropagateAtLaunch": true,
            "Value": {
              "Ref": "LoggingStreamName",
            },
          },
          {
            "Key": "Name",
            "PropagateAtLaunch": true,
            "Value": "Newsletters-TEST/AutoScalingGroupNewsletters",
          },
          {
            "Key": "Stack",
            "PropagateAtLaunch": true,
            "Value": "newsletters",
          },
          {
            "Key": "Stage",
            "PropagateAtLaunch": true,
            "Value": "TEST",
          },
        ],
        "TargetGroupARNs": [
          {
            "Ref": "TargetGroupNewsletters7256B07E",
          },
        ],
        "VPCZoneIdentifier": {
          "Ref": "newslettersPrivateSubnets",
        },
      },
      "Type": "AWS::AutoScaling::AutoScalingGroup",
    },
    "AutoScalingGroupNewslettersInstanceProfileF68637F3": {
      "Properties": {
        "Roles": [
          {
            "Ref": "InstanceRoleNewsletters073D61C6",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "AutoScalingGroupNewslettersLaunchConfigE2413199": {
      "DependsOn": [
        "InstanceRoleNewsletters073D61C6",
      ],
      "Properties": {
        "IamInstanceProfile": {
          "Ref": "AutoScalingGroupNewslettersInstanceProfileF68637F3",
        },
        "ImageId": {
          "Ref": "AMINewsletters",
        },
        "InstanceType": "t4g.small",
        "MetadataOptions": {
          "HttpTokens": "required",
        },
        "SecurityGroups": [
          {
            "Fn::GetAtt": [
              "GuHttpsEgressSecurityGroupNewsletters0D4F41FF",
              "GroupId",
            ],
          },
          {
            "Fn::GetAtt": [
              "WazuhSecurityGroup",
              "GroupId",
            ],
          },
        ],
        "UserData": {
          "Fn::Base64": {
            "Fn::Join": [
              "",
              [
                "#!/bin/bash
set -e
set +x
aws s3 cp s3://",
                {
                  "Ref": "DistributionBucketName",
                },
                "/newsletters/TEST/newsletters /tmp/app
chown -R ubuntu /tmp/app
su ubuntu -c '/usr/local/node/pm2 start --name newsletters /app/tmp/index.cjs'",
              ],
            ],
          },
        },
      },
      "Type": "AWS::AutoScaling::LaunchConfiguration",
    },
    "CertificateNewsletters752D28AD": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "DomainName": "newsletters.test.dev-gutools.co.uk",
        "Tags": [
          {
            "Key": "App",
            "Value": "newsletters",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/newsletters-nx",
          },
          {
            "Key": "Name",
            "Value": "Newsletters-TEST/CertificateNewsletters",
          },
          {
            "Key": "Stack",
            "Value": "newsletters",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "ValidationMethod": "DNS",
      },
      "Type": "AWS::CertificateManager::Certificate",
      "UpdateReplacePolicy": "Retain",
    },
    "DescribeEC2PolicyFF5F9295": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "autoscaling:DescribeAutoScalingInstances",
                "autoscaling:DescribeAutoScalingGroups",
                "ec2:DescribeTags",
                "ec2:DescribeInstances",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "describe-ec2-policy",
        "Roles": [
          {
            "Ref": "InstanceRoleNewsletters073D61C6",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "GetDistributablePolicyNewslettersDC325273": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:s3:::",
                    {
                      "Ref": "DistributionBucketName",
                    },
                    "/newsletters/TEST/newsletters/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GetDistributablePolicyNewslettersDC325273",
        "Roles": [
          {
            "Ref": "InstanceRoleNewsletters073D61C6",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "GuHttpsEgressSecurityGroupNewsletters0D4F41FF": {
      "Properties": {
        "GroupDescription": "Allow all outbound HTTPS traffic",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound HTTPS traffic",
            "FromPort": 443,
            "IpProtocol": "tcp",
            "ToPort": 443,
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "newsletters",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/newsletters-nx",
          },
          {
            "Key": "Stack",
            "Value": "newsletters",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VpcId": {
          "Ref": "VpcId",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "GuHttpsEgressSecurityGroupNewslettersfromNewslettersTESTLoadBalancerNewslettersSecurityGroup79A079AA300040353A3C": {
      "Properties": {
        "Description": "Load balancer to target",
        "FromPort": 3000,
        "GroupId": {
          "Fn::GetAtt": [
            "GuHttpsEgressSecurityGroupNewsletters0D4F41FF",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": {
          "Fn::GetAtt": [
            "LoadBalancerNewslettersSecurityGroup9C95445B",
            "GroupId",
          ],
        },
        "ToPort": 3000,
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
    "GuLogShippingPolicy981BFE5A": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "kinesis:Describe*",
                "kinesis:Put*",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:kinesis:eu-west-1:",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":stream/",
                    {
                      "Ref": "LoggingStreamName",
                    },
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GuLogShippingPolicy981BFE5A",
        "Roles": [
          {
            "Ref": "InstanceRoleNewsletters073D61C6",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "InstanceRoleNewsletters073D61C6": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "ec2.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AmazonSSMManagedInstanceCore",
              ],
            ],
          },
        ],
        "Path": "/",
        "Tags": [
          {
            "Key": "App",
            "Value": "newsletters",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/newsletters-nx",
          },
          {
            "Key": "Stack",
            "Value": "newsletters",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ListenerNewsletters17AACCE9": {
      "Properties": {
        "Certificates": [
          {
            "CertificateArn": {
              "Ref": "CertificateNewsletters752D28AD",
            },
          },
        ],
        "DefaultActions": [
          {
            "AuthenticateOidcConfig": {
              "AuthenticationRequestExtraParams": {
                "hd": "guardian.co.uk",
              },
              "AuthorizationEndpoint": "https://accounts.google.com/o/oauth2/v2/auth",
              "ClientId": {
                "Ref": "ClientId",
              },
              "ClientSecret": "{{resolve:secretsmanager:/TEST/deploy/newsletters/clientSecret:SecretString:::}}",
              "Issuer": "https://accounts.google.com",
              "OnUnauthenticatedRequest": "authenticate",
              "Scope": "openid",
              "TokenEndpoint": "https://oauth2.googleapis.com/token",
              "UserInfoEndpoint": "https://openidconnect.googleapis.com/v1/userinfo",
            },
            "Order": 1,
            "Type": "authenticate-oidc",
          },
          {
            "Order": 2,
            "TargetGroupArn": {
              "Ref": "TargetGroupNewsletters7256B07E",
            },
            "Type": "forward",
          },
        ],
        "LoadBalancerArn": {
          "Ref": "LoadBalancerNewsletters7869FC55",
        },
        "Port": 443,
        "Protocol": "HTTPS",
      },
      "Type": "AWS::ElasticLoadBalancingV2::Listener",
    },
    "LoadBalancerNewsletters7869FC55": {
      "Properties": {
        "LoadBalancerAttributes": [
          {
            "Key": "deletion_protection.enabled",
            "Value": "true",
          },
        ],
        "Scheme": "internet-facing",
        "SecurityGroups": [
          {
            "Fn::GetAtt": [
              "LoadBalancerNewslettersSecurityGroup9C95445B",
              "GroupId",
            ],
          },
        ],
        "Subnets": {
          "Ref": "newslettersPublicSubnets",
        },
        "Tags": [
          {
            "Key": "App",
            "Value": "newsletters",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/newsletters-nx",
          },
          {
            "Key": "Stack",
            "Value": "newsletters",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Type": "application",
      },
      "Type": "AWS::ElasticLoadBalancingV2::LoadBalancer",
    },
    "LoadBalancerNewslettersSecurityGroup9C95445B": {
      "Properties": {
        "GroupDescription": "Automatically created Security Group for ELB NewslettersTESTLoadBalancerNewsletters6CAE8321",
        "SecurityGroupIngress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow from anyone on port 443",
            "FromPort": 443,
            "IpProtocol": "tcp",
            "ToPort": 443,
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "newsletters",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/newsletters-nx",
          },
          {
            "Key": "Stack",
            "Value": "newsletters",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VpcId": {
          "Ref": "VpcId",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "LoadBalancerNewslettersSecurityGrouptoNewslettersTESTGuHttpsEgressSecurityGroupNewslettersF696E9FB300031CB4500": {
      "Properties": {
        "Description": "Load balancer to target",
        "DestinationSecurityGroupId": {
          "Fn::GetAtt": [
            "GuHttpsEgressSecurityGroupNewsletters0D4F41FF",
            "GroupId",
          ],
        },
        "FromPort": 3000,
        "GroupId": {
          "Fn::GetAtt": [
            "LoadBalancerNewslettersSecurityGroup9C95445B",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "ToPort": 3000,
      },
      "Type": "AWS::EC2::SecurityGroupEgress",
    },
    "NewslettersCname": {
      "Properties": {
        "Name": "newsletters.test.dev-gutools.co.uk",
        "RecordType": "CNAME",
        "ResourceRecords": [
          {
            "Fn::GetAtt": [
              "LoadBalancerNewsletters7869FC55",
              "DNSName",
            ],
          },
        ],
        "Stage": "TEST",
        "TTL": 3600,
      },
      "Type": "Guardian::DNS::RecordSet",
    },
    "ParameterStoreReadNewslettersD7903C98": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:eu-west-1:",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/newsletters/newsletters",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:eu-west-1:",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/newsletters/newsletters/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "parameter-store-read-policy",
        "Roles": [
          {
            "Ref": "InstanceRoleNewsletters073D61C6",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TargetGroupNewsletters7256B07E": {
      "Properties": {
        "HealthCheckIntervalSeconds": 10,
        "HealthCheckPath": "/healthcheck",
        "HealthCheckProtocol": "HTTP",
        "HealthCheckTimeoutSeconds": 5,
        "HealthyThresholdCount": 5,
        "Port": 3000,
        "Protocol": "HTTP",
        "Tags": [
          {
            "Key": "App",
            "Value": "newsletters",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/newsletters-nx",
          },
          {
            "Key": "Stack",
            "Value": "newsletters",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "TargetGroupAttributes": [
          {
            "Key": "deregistration_delay.timeout_seconds",
            "Value": "30",
          },
          {
            "Key": "stickiness.enabled",
            "Value": "false",
          },
        ],
        "TargetType": "instance",
        "UnhealthyThresholdCount": 2,
        "VpcId": {
          "Ref": "VpcId",
        },
      },
      "Type": "AWS::ElasticLoadBalancingV2::TargetGroup",
    },
    "WazuhSecurityGroup": {
      "Properties": {
        "GroupDescription": "Allow outbound traffic from wazuh agent to manager",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Wazuh event logging",
            "FromPort": 1514,
            "IpProtocol": "tcp",
            "ToPort": 1514,
          },
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Wazuh agent registration",
            "FromPort": 1515,
            "IpProtocol": "tcp",
            "ToPort": 1515,
          },
        ],
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/newsletters-nx",
          },
          {
            "Key": "Stack",
            "Value": "newsletters",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VpcId": {
          "Ref": "VpcId",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
  },
}
`;
